### PR TITLE
Only do 1 token update a time

### LINKED ISSUE #7062: Multiple external auth attempts occur simultaneously, making the first get lost
## Checklist

- [x] I have updated to the latest available Home Assistant version.
- [x] I have cleared the cache of my browser.
- [x] I have tried a different browser to see if it is related to my browser.

## The problem
Occasionally, the frontend asks for authentication 2 or more times on the external bus, e.g.:

```
2020-09-18 21:36:39.172 [Verbose] [main] [WebViewController.swift:644] userContentController(_:didReceive:) > getExternalAuth called, forced: false
2020-09-18 21:36:39.229 [Verbose] [main] [WebViewController.swift:644] userContentController(_:didReceive:) > getExternalAuth called, forced: false
```

and the app dutifully responds to both:

```
2020-09-18 21:36:39.247 [Verbose] [main] [WebViewController.swift:656] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-09-18 21:36:39.247 [Verbose] [main] [WebViewController.swift:656] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
```

however, the frontend spins with a "reconnecting…" method for a few seconds. I believe this is because whatever asks for an authentication token first gets stomped because the callbacks are constant string-based keys:

https://github.com/home-assistant/frontend/blob/82f80db558ceaab1d1cbe54fbfe2e2165e7cacd5/src/external_app/external_auth.ts#L77-L79

## Expected behavior
Either the frontend doesn't call the external bus multiple times simultaneously, or it handles all that it does without losing the promises.

## Steps to reproduce
This requires a certain ordering of network requests, but it seems to be made worse by having cards that will cause a network request in addition to the WebSocket connection. Backgrounding the iOS app for a while until the token expires is a good way to reproduce it.

## Environment

- Home Assistant release with the issue: core @ 62c4e072f5e21ccd057e343c700d2f81a0a50b2e
- Last working Home Assistant release (if known): unknown
- Browser and browser version: iOS app throughout the ages, so Safari-ish
- Operating system: iOS 13

## Additional information

I may be completely misreading the JavaScript.

### PR DESCRIPTION
**Pull request by bramkragten**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes https://github.com/home-assistant/frontend/issues/7062
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/external_app/external_auth.ts [hunk 1]
```diff
@@ -64,15 +64,21 @@ export class ExternalAuth extends Auth {
     });
   }
 
+  private _tokenCallbackPromise?: Promise<RefreshTokenResponse>;
+
   public async refreshAccessToken(force?: boolean) {
+    if (this._tokenCallbackPromise) {
```
Comment (ludeeus) [inline src/external_app/external_auth.ts:70]:
```suggestion
    if (this._tokenCallbackPromise && !force) {
```

We should create a new if  `force` is passed right?

**Review by ludeeus — APPROVED** [review]

**Comment by zacwest** [comment]
If multiple happen and force is true in any of them, won't that cause the same issue where the external bus call itself is only able to track one at once?

**Comment by bramkragten** [comment]
Why would there be multiple force requests? I think it is only used when loading the app and hitting a wrong token?

**Comment by zacwest** [comment]
The problem scenario I'm imagining is something like…

1. refresh, force=false
2. refresh, force=false
3. refresh, force=true

would cause 1 and 2 to be lost because of how the callback works in the external bus, if I am understanding _that_ side of things correctly.

**Comment by bramkragten** [comment]
I don't think that scenario will ever happen, I'll check the code in a sec

**Comment by bramkragten** [comment]
We only do a force when we load the core and get an invalid auth. The force parameter should trigger the reauth flow in the app, as if it was not authenticated. Since we are not connected at that point, it is not possible that a `force=false` request is made.

https://github.com/home-assistant/frontend/blob/dev/src/entrypoints/core.ts#L66

**Comment by zacwest** [comment]
My read of that (and I think what I have seen) is that we may use an invalid-due-to-expired auth token and that triggers the forced flow, so it's not just logout events.

**Comment by bramkragten** [comment]
this is not a logout event, it is when the apps starts, and the auth token is no longer valid (removed?).

When the app starts we connect to HA and try to get the access token (without force) from the app, if that fails with an invalid auth error, we will call refresh access token with force. That means all previous promises can be discarded as we will re-do the entire auth (login page). After the re-login of the user, we will try to create the connection again.

**Comment by chriss158** [comment]
Every now and then i can't get a reconnection to my home assistant after the connection is broken (ex. duo to a home assistant restart) with the android app. The frontend just shows the message "Connection lost. Reconnecting..." for ever. I had never problems with release before this PR.

Did some debugging the last couple of days and tracked down the problem to this PR. 

I think this is reproducible with following steps:

1. Deactivate "Close connection automatically" in the profile settings of HA.
2. Open the android app and let it connect to home assistant
3. Lock the phone/tablet leaving the app open and in foreground
4. Wait around 30 minutes (I think this is the session timeout for the token)
5. Restart HA and wait till it is running fully again
5. Unlock the phone/tablet.

While HA is restarting the frontend tries to reconnect in background.

1. `RefreshAccessToken`is called 
https://github.com/home-assistant/frontend/blob/4973d8f629c3619f7e113b6fb6888aaf3027ede6/src/external_app/external_auth.ts#L69

2.  `RefreshAccessToken`calls `getExternalAuth`
https://github.com/home-assistant/frontend/blob/4973d8f629c3619f7e113b6fb6888aaf3027ede6/src/external_app/external_auth.ts#L92

3. `getExternalAuth` triggers the `onGetExternalAuth` of the android app 
https://github.com/home-assistant/android/blob/6e13b162144cfeed69d0d37f050098d2b16ebf6c/app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt#L82

4. The method call `view.setExternalAuth` will fail because Home assistant is not yet restarted.

5. The catch block is triggered and calls the following method:
https://github.com/home-assistant/android/blob/6e13b162144cfeed69d0d37f050098d2b16ebf6c/app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt#L89

6. Now the `_tokenCallbackPromise` is trigged 
https://github.com/home-assistant/frontend/blob/4973d8f629c3619f7e113b6fb6888aaf3027ede6/src/external_app/external_auth.ts#L81
with success = false. Meaning the _tokenCallbackPromise is in a rejected state.

7. Now after one second the next try for a reconnect starts.  
https://github.com/home-assistant/home-assistant-js-websocket/blob/1836b3ae57309a969627fc993619cfd40b47204c/lib/socket.ts#L84

8. Again `RefreshAccessToken` is called because of expired session token. This time from the reconnection instead of the app https://github.com/home-assistant/home-assistant-js-websocket/blob/1836b3ae57309a969627fc993619cfd40b47204c/lib/socket.ts#L87

9. Then this check is called:
https://github.com/home-assistant/frontend/blob/4973d8f629c3619f7e113b6fb6888aaf3027ede6/src/external_app/external_auth.ts#L70  `_tokenCallbackPromise` is not undefined and force is undefined because the parameter is not defined by the caller.  Then the `_tokenCallbackPromise` is called. And here is the **culprint**: `tokenCallbackPromise` is in a rejected state with undefined data.   The rejected promise triggers an exception.

10. The exception is then handled here:
https://github.com/home-assistant/home-assistant-js-websocket/blob/1836b3ae57309a969627fc993619cfd40b47204c/lib/socket.ts#L90
This triggers that the websocket will be closed again.

11. Then the reconnection process is beginning from the start. And we are in a reconnect loop.

On the home assistant side there are following messages. Which explains this process above.

```
2020-11-16 17:31:00 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection.140600942755264] Connected from ipaddress
2020-11-16 17:31:00 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection.140600942755264] Sending {'type': 'auth_required', 'ha_version': '0.117.6'}
2020-11-16 17:31:00 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection.140600942755264] Disconnected
```

### Long story short
Maybe here
https://github.com/home-assistant/frontend/blob/4973d8f629c3619f7e113b6fb6888aaf3027ede6/src/external_app/external_auth.ts#L70

should be a check if the promise is rejected. If it is reject then do not return the reject promise. Instead do the stuff below (call getExternalAuth etc.) If the promise is resolved, then return the promise.

I hope i could describe the problem understandable. 

Best regards
Chris

**Comment by bramkragten** [comment]
Hi Chris, thanks for your lengthy tracking, but could you create an issue for this instead of commenting on a merged PR? Thanks!
